woocommerce/woocommerce
# Support testing many variants of a block at runtime via dynamically generated templates

### Changes proposed in this Pull Request:

This introduces a system in e2e tests for blocks that utilizes handlebars templates and wp-cli. It allows you to, at runtime, generate post content containing your block under test with many different attribute scenarios.

In this I reimplemented the e2e tests for stock status filter that I introduced in https://github.com/woocommerce/woocommerce/pull/44212

When considering all the things we could do, I didn't want to upfront generate lots of variations or anything like that. What we really want is a simple way to generate the block with dynamic attributes at runtime.  This is extremely flexible and with Handlebars being a flexible templating language, the things you can do are almost unlimited. In future though we could precompile templates or create some other pre-test system using the same concepts.

To get us started I added a simple helper that lets you render a wordpress block in a template and pass it attributes.

As you can see in my example template it looks like this:

```
{{#> wp-block blockName='woocommerce/product-filter-stock-status' attributes=attributes }}
{{/ wp-block }}
```

This also supports rendering child content in the block if you want to. 

This example only scratches the surface, you could use loops or other Handlebars features to generate virtually any template that suits you.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. See the CI pass
2. You could also try checking this out exploring locally, adding some templates of your own.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

Add a system to support dynamic template testing in blocks e2e tests.

</details>
